crates/shared/src/coordinator/mod.rs

```diff
@@ -0,0 +1,682 @@
+    /// race.
+    ///
+    /// The primary cause of this has to due with the interface of the
+    /// [`BuilderStateCoordinator`]'s API.  In general, we want to be able to retrieve
+    /// a [`BuilderState`] via the [`BuilderStateId`]. The [`BuilderStateId`] only references
+    /// a [`ViewNumber`](hotshot_types::data::ViewNumber) and a [`VidCommitment`](`hotshot_types::vid::VidCommitment`).
+    /// While this information is available in the [`QuorumProposal`],
+    /// it only helps us to rule out [`BuilderState`]s that already exist.
+    /// It does **NOT** help us to pick a [`BuilderState`] that is the best fit to extend from.
+    ///
+    /// This is where the `justify_qc` comes in to consideration.  The `justify_qc`
+    /// contains the previous [`ViewNumber`](hotshot_types::data::ViewNumber) that is
+    /// being extended from, and in addition it also contains the previous
+    /// [`Commitment<Leaf<Types>>`](`committable::Commitment`)
+    /// that is being built on top of.  Since our [`BuilderState`]s store identifying
+    /// information that contains this same `leaf_commit` we can compare these
+    /// directly to ensure that we are extending from the correct [`BuilderState`].
+    ///
+    /// This function determines the best [`BuilderState`] in the following steps:
+    ///
+    /// 1. If we have a [`BuilderState`] that is already spawned for the current
+    ///    [`QuorumProposal`], then we should should return no states, as one already
+    ///    exists.  This will prevent us from attempting to spawn duplicate
+    ///    [`BuilderState`]s.
+    /// 2. Attempt to find all [`BuilderState`]s that are recorded within
+    ///    coordinator that have matching view number and leaf commitments. There
+    ///    *should* only be one of these.  But all would be valid extension points.
+    /// 3. If we can't find any [`BuilderState`]s that match the view number
+    ///    and leaf commitment, then we should return for the maximum stored view
+    ///    number that is smaller than the current [`QuorumProposal`].
+    /// 4. If there is is only one [`BuilderState`] stored in the coordinator, then
+    ///    we should return that [`BuilderState`] as the best fit.
+    /// 5. If none of the other criteria match, we return an empty result as it is
+    ///    unclear what to do in this case.
+    ///
+    /// <div class="warning">
+    ///
+    ///  Note: Any time this function returns more than a single entry in its
+    ///  result, there is a potential for a race condition.  This is
+    ///  because there are multiple [`BuilderState`]s that are equally valid to
+    ///  extend from.  This race could be avoided by just picking one of the
+    ///  entries in the resulting [`Vec`], but this is not done here in order
+    ///  to allow us to highlight the possibility of the race.
+    ///
+    /// </div>
+    #[must_use]
+    async fn find_builder_states_to_extend(
```

Comment:
Is this the same as the original `best_builder_states_to_extend` function or is there any logic change? I think we should also use the spawned builder commitments to determine the best states, as written in the design doc [here](https://www.notion.so/espressosys/Builder-State-Refactor-1022431b68e9809bb6dfe54346aa0aef?pvs=4#10a2431b68e980b69f1ef9a1edcae230).

Reply:
I think the builder commitment in the proposal corresponds to the builder state that built this block originally. So it's like:
* We extend a builder state to build a block and store the new builder commitment in the `builder_commitments` of this builder state.
* We send the block to HotShot.
* HotShot uses the block to form a proposal.
* We get the proposal, which also has a builder commitment.
* We look up the builder commitment in all `builder_commitments`.

I may be wrong! Let me know if so.

Reply:
Oh, I see! Sorry if this was already discussed before, but I struggle to imagine situation where we can't find correct builder state by leaf commitment but builder commitment matches

Reply:
Good point. I just found a relevant Zulip thread. Left a comment there: https://espresso.zulipchat.com/#narrow/channel/419378-eng-block-builder/topic/Proposal.20dependencies/near/482281714.